### PR TITLE
phi2 contrib ops changes

### PR DESCRIPTION
**Pull request by wangyems**
### Description

1. support causal mask in MHA cpu
2. support custom rotary_dim in rotary_emb
3. add bf16 for rotary_emb
4. fix a bug in attention rotary

### CODE REVIEW

File: onnxruntime/test/contrib_ops/rotary_embedding_op_test.cc [hunk 1]
```diff
@@ -143,13 +176,31 @@ static void RunTests(const std::vector<float>& input_data,
...
+    //         sin_cache,
+    //         output_data,
+    //         batch_size,
+    //         sequence_length,
+    //         head_size,
+    //         rotary_embedding_dim,
+    //         num_heads,
+    //         max_sequence_length,
+    //         interleaved,
+    //         TensorType::kBFloat16,
```
Comment (wangyems) [inline onnxruntime/test/contrib_ops/rotary_embedding_op_test.cc:200]: bf16 test has larger diff. temporarily comment it out.

File: onnxruntime/contrib_ops/cuda/bert/rotary_embedding_impl.cu [hunk 2]
```diff
@@ -82,20 +92,30 @@ Status LaunchRotaryEmbeddingKernel(
...
-  const dim3 block(head_size, 1, 1);
-
   // Note: Current implementation assumes head_size <= max_threads_per_block
   // because head_size is currently large for LLaMA-2. For smaller head_size
   // and num_heads values, we can create a block as `block(num_heads, head_size, 1)`
   // instead. This will require kernel changes to support.
+  ORT_ENFORCE(head_size <= max_threads_per_block,
+              "Rotary embedding dim must be <= max_threads_per_block");
+
+  int tpb = head_size;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/rotary_embedding_impl.cu:108]: What is this logic for?

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/rotary_embedding_impl.cu:108]: it is to ensure the block_size be multiple of warp size.

Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/bert/rotary_embedding_impl.cu:108]: why not just (head_size + 31)/32*32 work?

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/rotary_embedding_impl.cu:108]: it uses fewer instructions but both work

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/rotary_embedding_impl.cu]: changed to (head_size + 31)/32*32

Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/bert/rotary_embedding_impl.cu]: it doesn't actually use less instructions. (head_size + 31)/32*32 uses 1 add, 1 right shift and 1 left shift, in total, 3 instructions.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Comment by snnn** [comment]
This PR has been cherry-picked into the `rel-1.17.0` branch in PR #19243. Removing the `release:1.17.0` label.

**Pull request merged**